PT-2340: support MySQL 8.4

svetasmirnova opened the pull request:
This huge PR adds MySQL 8.4 support to the Percona Toolkit.

- [x] The contributed code is licensed under GPL v2.0
- [x] Contributor Licence Agreement (CLA) is signed
- [x] util/update-modules has been ran
- [x] Documentation updated
- [x] Test suite update

svetasmirnova committed:
PT-2340 - Support MySQL 8.4

- Adjusted sandbox scripts, so they can start MySQL 8.4
- Added MySQL 8.4 configuration file
- Removed empty file sandbox/5.6
- Removed unused files sandbox/set-mysql, sandbox/slave_channels_t.sql, sandbox/jenkins-test
- Removed offensive terminology from the sandbox scripts wherever it is possible

svetasmirnova committed:
PT-2340 - Support MySQL 8.4

- Fixed status check for 8.4 sandbox
- Removed option "reset" from the usage output, because option itself was removed by 92a36715c44663db82e2cf5e78b98b622e905727
- Fixed sandbox/start-sandbox script, so it can start custom setups with 8.4

svetasmirnova committed:
PT-2340 - Support MySQL 8.4

- Removed offensive terminology from library files and their tests
- Removed unused sandbox/prove2junit.pl
- Added option mysql_ssl to DSN and possibility to have DSN of multiple letters

svetasmirnova committed:
PT-2340 - Support MySQL 8.4

- Removed all unneeded occuriences of the word "master" in lib

svetasmirnova committed:
PT-2340 - Support MySQL 8.4

- Removed lib/Percona/Test.pm, lib/Safeguards.pm, t/lib/Safeguards.t, because they are not used anymore
- Removed word "slave" from lib

svetasmirnova committed:
PT-2340 - Support MySQL 8.4

- Removed runtime.txt after discussion with Anastasia Alexandrova
- Added "use VersionParser" into tests in t/lib when needed
- Removed word master from tests for pt-archiver, pt-config-diff, pt-deadlock-logger, pt-duplicate-key-checker, pt-find, pt-fk-error-logger, pt-heartbeat, pt-index-usage, pt-ioprofile, pt-kill, pt-mysql-summary
- Removed word slave from tests for pt-archiver, pt-config-diff, pt-deadlock-logger, pt-duplicate-key-checker, pt-find, pt-fk-error-logger, pt-heartbeat, pt-index-usage, pt-ioprofile, pt-kill, pt-mysql-summary
- Updated modules for pt-archiver, pt-config-diff, pt-deadlock-logger, pt-duplicate-key-checker, pt-find, pt-fk-error-logger, pt-heartbeat, pt-index-usage, pt-ioprofile, pt-kill, pt-mysql-summary
- Changed mysql_ssl patch, so it is now short option s
- Added a check for existing zombies in t/pt-kill/execute_command.t
- Added bin/pt-galera-log-explainer to .gitignore

svetasmirnova committed:
PT-2340 - Support MySQL 8.4

- Updated modules and tests for pt-online-schema-change
- Removed typo from lib/MasterSlave.pm

svetasmirnova committed:
PT-2340 - Support MySQL 8.4

- Updated modules and tests for pt-query-digest, pt-show-grants, pt-slave-delay, pt-slave-find, pt-slave-restart, pt-stalk, pt-summary

svetasmirnova committed:
PT-2340 - Support MySQL 8.4

- Updated modules and tests for pt-query-digest, pt-show-grants, pt-slave-delay, pt-slave-find, pt-slave-restart, pt-stalk, pt-summary, pt-table-checksum

svetasmirnova committed:
Merge branch 'PT-2340_Support_MySQL_8.4' of github.com:percona/percona-toolkit into PT-2340_Support_MySQL_8.4

svetasmirnova committed:
PT-2340 - Support MySQL 8.4

- Updated modules and tests for pt-table-sync, pt-table-usage, pt-upgrade, pt-variable-advisor

svetasmirnova committed:
PT-2340 - Support MySQL 8.4

- Fixed staff I broke for 8.0

svetasmirnova committed:
PT-2340 - Support MySQL 8.4

- Fixed 5.7 tests

svetasmirnova committed:
PT-2340 - Support MySQL 8.4

- Added deprecation warning to pt-slave-delay
- Re-enabled tests for pt-slave-delay to check the warning
- Disabled pt-slave-delay for MySQL 8.1+

svetasmirnova committed:
PT-2340 - Support MySQL 8.4

- Created pt-replica-restart. pt-slave-restart kept as symbolic link

svetasmirnova committed:
PT-2340 - Support MySQL 8.4

- Changed write-user-docs script, so it creates special documentation pages for symbolic links.

svetasmirnova committed:
PT-2340 - Support MySQL 8.4

- Fixed comparison operators for VersionCheck calls

svetasmirnova committed:
PT-2340 - Support MySQL 8.4

- Created pt-replica-find. pt-slave-find kept as symbolic link

svetasmirnova committed:
PT-2340 - Support MySQL 8.4

- Updated documentation

svetasmirnova committed:
PT-2340 - Support MySQL 8.4

- Fixed typos
- Added tests for MyRocks and TokuDB clustered indexes into t/pt-duplicate-key-checker/clustered_keys.t
- Updated modules

svetasmirnova committed:
Merge branch '3.x' into PT-2340_Support_MySQL_8.4

svetasmirnova committed:
PT-2340 - Support MySQL 8.4

- Updated tests for pt-variable-advisor and lib/VariableAdvisorRules

svetasmirnova committed:
PT-2340 - Support MySQL 8.4

- Moved data collection for THP from lib/bash/report_system_info.sh to lib/bash/collect_system_info.sh, so pt-summary reports THP status on the machine where samples were collected, not on the machine where an engineer examines samples.

svetasmirnova committed:
PT-2340 - Support MySQL 8.4

- Adjusted test results, broken due to mysql.component table in PS

svetasmirnova committed:
PT-2340 - Support MySQL 8.4

- Tests for the minimal SSL support
- Updated util/update-modules, so they don't skip tools with modules, not defined in lib

svetasmirnova committed:
Merge branch '3.x' into PT-2340_Support_MySQL_8.4

svetasmirnova committed:
PT-2340 - Support MySQL 8.4

- Added tests for legacy style heartbeat and checksum tables

svetasmirnova committed:
PT-2340 - Support MySQL 8.4

- Added test for replica lag check for pt-archiver
- Re-added deprecated slave- options
- Added tests for deprecation warnings and for legacy options for pt-archiver
- Removed practically not supported options --replica-user and --replica-password from pt-archiver, pt-kill, pt-query-digest
- Added legacy source/replica options (master/slave) variants and tests for pt-heartbeat, pt-online-schema-change, pt-replica-find, pt-replica-restart, pt-table-checksum, pt-table-sync
- Updated modules after lib/MasterSlave.pm change

svetasmirnova committed:
PT-2340 - Support MySQL 8.4

- Adjusted t/pt-deadlock-logger/check_schema_exists.t so it makes sense

svetasmirnova committed:
PT-2340 - Support MySQL 8.4

- Adjusted pt-variable-advisor, so it reflects current defaults

svetasmirnova committed:
PT-2340 - Support MySQL 8.4

- Updated modules

svetasmirnova committed:
PT-2340 - Support MySQL 8.4

- Fixed typo in bin/pt-table-sync

svetasmirnova committed:
Merge branch '3.x' into PT-2340_Support_MySQL_8.4

svetasmirnova committed:
PT-2340 - Support MySQL 8.4

- Fixed typos

ylacancellera commented on t/lib/bash/collect_mysql_info.sh:
```diff
@@ -163,11 +163,18 @@ cmd_ok \
+   mysql_version="$(get_var version "$p/mysql-variables")"
+   source_logs_file="mysql-binary-logs"
+   source_status_file="mysql-binary-log-status"
+   if [ "$mysql_version" '<' "8.1" ]; then
```
Is that planned for any mariadb version ? 
```
$ if [ "10.4" '<' "8.1" ]; then echo smaller; fi
smaller
```

svetasmirnova replied:
Good point! We do not care about MariaDB in our tests atm but we will in future versions. Could you please file a Jira ticket, so I can plan a fix for it?

ylacancellera replied:
https://perconadev.atlassian.net/browse/PT-2390

ylacancellera commented on t/pt-archiver/archive_using_channels.t:
```diff
@@ -17,67 +17,71 @@ require "$trunk/bin/pt-archiver";
-$sb->load_file('chan_master1', "sandbox/gtid_on.sql", undef, no_wait => 1);
-$sb->load_file('chan_master2', "sandbox/gtid_on.sql", undef, no_wait => 1);
-$sb->load_file('chan_slave1', "sandbox/slave_channels.sql", undef, no_wait => 1);
+if ( $sandbox_version lt '8.1' ) {
```
commenting again for -lt, this time on perl, just in case. there's MySQL 9 coming, I imagine MySQL 10 would exist at a point.
```
$ perl -e "if ( '10.0' lt '8.1') { print('smaller')}"
smaller
```

svetasmirnova replied:
Good point! We do not care about MariaDB in our tests atm but we will in future versions. Could you please file a Jira ticket, so I can plan a fix for it?

ylacancellera replied:
https://perconadev.atlassian.net/browse/PT-2390

ylacancellera commented on bin/pt-archiver:
```diff
@@ -7179,14 +7292,14 @@ sub main {
-               my $lag = $ms->get_slave_lag($lag_dbh);
+               my $lag = $ms->get_replica_lag($lag_dbh);
                while ( !defined $lag || $lag > $o->get('max-lag') ) {
                   PTDEBUG && _d("Sleeping: slave lag for server '$id' is", $lag);
```
Since pt-heartbeat had his schema changed to reflect the source/replica change, does this one should be changed for coherence ?
Or, is it left as is so to keep the current output ?

svetasmirnova replied:
Hi! I don't understand your comment. Modified code uses get_replica_lag. What needs to be changed?

ylacancellera replied:
It is for `_d("Sleeping: slave lag for server '$id' is", $lag);`, on row 7297 and 7299 (on new branch).
7297 is with PTDEBUG, so really not a problem, but the other one is a log to users
"replica lag" would make it more coherent, but that's only a nitpick

svetasmirnova replied:
Fixed, please review again

ylacancellera replied:
lgtm for the bin, the tests will need some fixing as well
```
$ grep -rn 'Sleeping:'
t/pt-archiver/replica_lag.t:73:   qr/Sleeping: slave lag for server/,
t/pt-archiver/replica_lag.t:101:   qr/Sleeping: slave lag for server/,
t/pt-archiver/replica_lag.t:129:   qr/Sleeping: slave lag for server/,
t/pt-archiver/replica_lag.t:156:   qr/Sleeping: slave lag for server/,
t/pt-archiver/replica_lag.t:184:   qr/Sleeping: slave lag for server/,
t/pt-archiver/replica_lag.t:212:   qr/Sleeping: slave lag for server/,
t/pt-archiver/replica_lag.t:240:   qr/Sleeping: slave lag for server/,
bin/pt-archiver:7297:                  PTDEBUG && _d("Sleeping: replica lag for server '$id' is", $lag);
bin/pt-archiver:7299:                     _d("Sleeping: replica lag for server '$id' is", $lag);
```

svetasmirnova replied:
Fixed

ylacancellera commented on t/pt-heartbeat/multi_update_mode.t:
```diff
@@ -60,175 +61,231 @@ foreach my $port (@ports) {
-   $rows->{12346}->{relay_master_log_file} && $rows->{12346}->{exec_master_log_pos},
-   "Slave1 relay_master_log_file and exec_master_log_pos for master"
-);
+   $rows->{12346}->{"relay_source_log_file"} && $rows->{12346}->{"exec_source_log_pos"},
```
info: why is this test (and next one) not using `relay_$(source_name}_log_file` like the previous one ? Can it miss an edge case ?
"test" as in `ok(some conditions)` not as in test file

svetasmirnova replied:
Because it selects from the test.heartbeat table, and not from SHOW SLAVE/REPLICA STATUS. This table has static column names, not depending on server version. Legacy column names are supported, this is tested by different test: see t/pt-heartbeat/heartbeat_table.t

ylacancellera replied:
Got it thank you

svetasmirnova committed:
PT-2340 - Support MySQL 8.4

- s/slave lag/replica lag/ in reports

ylacancellera commented on lib/MySQLProtocolParser.pm:
```diff
@@ -75,7 +75,7 @@ use constant {
    COM_TABLE_DUMP          => '13',
    COM_CONNECT_OUT         => '14',
-   COM_REGISTER_SLAVE      => '15',
+   COM_REGISTER_REPLICA    => '15',
```
The packet exists on 8.0 and 8.4, but not 5.7
I don't find any code usage of it in any lib/bin though

svetasmirnova replied:
Changed back to COM_REGISTER_SLAVE , because this is still the name of the constant in 8.4.

ylacancellera commented on lib/PerconaTest.pm:
```diff
@@ -78,6 +83,19 @@ eval {
+our $source_reset = 'binary logs and gtids';
+our $source_change = 'replication source';
+our $replica_name = 'replica';
+if ( $sandbox_version lt '8.1' || ( $ENV{FORK} || "" eq 'mariadb' ) ) {
```
works in an unintended way (PT-2390). I listed this code path in PT-2390 for review. I'm still commenting since it's a different condition
10.x will return true on `$sandbox_version lt '8.1'`, so the second part on the fork is not used.
```
$ echo $sandbox_version $FORK
10.4 mysql
$ perl -e "if ( $sandbox_version lt '8.1' || ( $ENV{FORK} || '' eq 'mariadb' ) ) { print('works') }"
works
```
As long as mysql 10 does not exist it still does what we want

svetasmirnova replied:
Will ignore it for now, fix later.

ylacancellera commented on bin/pt-replica-find:
```diff
@@ -0,0 +1,4819 @@
+# Check at the end of this package for the call to main() which actually runs
+# the program.
+# ###########################################################################
+package pt_slave_find;
```
can this impact maintenance ? The names are mismatched between the toolname and the package name
If it's safe, good to me. I don't find any direct usage

svetasmirnova replied:
Fixed

ylacancellera commented on bin/pt-replica-find:
```diff
@@ -0,0 +1,4819 @@
+#    +- 1
+#       +- 2
+#    +- 1
+sub print_slaves {
```
Since many internal variables were renamed in lib, I'm commenting to be thorough
This tool package did not receive the s/slave/replica
However I don't see any misusage

Unless this was a goal to remove all of them in this PR specifically, I'm ok to keep it and maybe flag it for later

svetasmirnova replied:
fixed

ylacancellera commented on bin/pt-replica-find:
```diff
@@ -0,0 +1,4819 @@
+   }
+   push @lines, [
+      'Replication',
+      ($ss ? "Is a slave, " : "Is not a slave, ")
```
Unaligned with rest, I can flag it for later if needed

svetasmirnova replied:
Fixed

ylacancellera commented on bin/pt-replica-find:
```diff
@@ -0,0 +1,4819 @@
+     Version         5.1.34-log
+     Server ID       12346
+     Uptime          04:54 (started 2010-06-17T11:21:24)
+     Replication     Is a slave, has 1 slaves connected
```
Unaligned with the rest, but coherent with the current states of things
Again, I can flag it for later, or skip

svetasmirnova replied:
Fixed

ylacancellera reviewed (approved):
I have commented what I found already, it's only nitpicks on some missed "slave" wordings. (except the version comparisons, but PT-2390 have been created so I exclude it).

Looks good to me

EvgeniyPatlan commented on bin/pt-archiver:
```diff
@@ -4161,117 +4205,157 @@ sub wait_for_master {
+   my $replica2_name = 'replica';
+   if ( $vp2 lt '8.1' || $vp2->flavor() =~ m/maria/ ) {
+      $replica2_name = 'slave';
+   }
```
Maybe we can rework this?

```
sub get_replica_name {
    my ($dbh) = @_;
    my $vp = VersionParser->new($dbh);
    return ($vp lt '8.1' || $vp->flavor() =~ m/maria/) ? 'slave' : 'replica';
}

my $replica1_name = get_replica_name($s1_dbh);
my $replica2_name = get_replica_name($s2_dbh);
```

svetasmirnova replied:
Implemented

EvgeniyPatlan commented on bin/pt-heartbeat:
```diff
@@ -622,117 +666,157 @@ sub wait_for_master {
+   if ( $vp2 lt '8.1' || $vp2->flavor() =~ m/maria/ ) {
+      $replica2_name = 'slave';
+   }
+
```
```
sub get_replica_name {
    my ($dbh) = @_;
    my $vp = VersionParser->new($dbh);
    return ($vp lt '8.1' || $vp->flavor() =~ m/maria/) ? 'slave' : 'replica';
}

my $replica1_name = get_replica_name($s1_dbh);
my $replica2_name = get_replica_name($s2_dbh);
```

EvgeniyPatlan commented on bin/pt-kill:
```diff
@@ -4438,117 +4482,157 @@ sub wait_for_master {
+   if ( $vp2 lt '8.1' || $vp2->flavor() =~ m/maria/ ) {
+      $replica2_name = 'slave';
+   }
+
```
```
sub get_replica_name {
    my ($dbh) = @_;
    my $vp = VersionParser->new($dbh);
    return ($vp lt '8.1' || $vp->flavor() =~ m/maria/) ? 'slave' : 'replica';
}

my $replica1_name = get_replica_name($s1_dbh);
my $replica2_name = get_replica_name($s2_dbh);
```

svetasmirnova replied:
Implemented

EvgeniyPatlan commented on bin/pt-query-digest:
```diff
@@ -11032,117 +11076,157 @@ sub wait_for_master {
+   my $replica2_name = 'replica';
+   if ( $vp2 lt '8.1' || $vp2->flavor() =~ m/maria/ ) {
+      $replica2_name = 'slave';
+   }
```
```

sub get_replica_name {
    my ($dbh) = @_;
    my $vp = VersionParser->new($dbh);
    return ($vp lt '8.1' || $vp->flavor() =~ m/maria/) ? 'slave' : 'replica';
}

my $replica1_name = get_replica_name($s1_dbh);
my $replica2_name = get_replica_name($s2_dbh);

```

svetasmirnova replied:
Implemented

svetasmirnova committed:
PT-2340 - Support MySQL 8.4

- s/slave lag/replica lag/ in tests
- reverted rename of COM_REGISTER_SLAVE to COM_REGISTER_REPLICA, because the constant still named COM_REGISTER_SLAVE in 8.4
- removed some slave leftovers from pt-replica-find

svetasmirnova committed:
PT-2340 - Support MySQL 8.4

- Implemented functions get_replica_name and get_source_name in lib/MasterSlave.pm I did not implement similar functions in other places, because they set other variables as well, not only replica_name or source_name.

svetasmirnova committed:
PT-2340 - Support MySQL 8.4

- Adjusted tools and tests which fail due to fixed update-modules command
- Adjsted t/pt-table-checksum/basics.t test for 8.0

svetasmirnova committed:
PT-2340 - Support MySQL 8.4

- Added SSL to 8.4 tests

svetasmirnova committed:
PT-2340 - Support MySQL 8.4

- Added sleep to t/lib/MasterSlave.t, so it does not fail on fast machines
- Disabled SSL tests on versions earlier than 8.0

svetasmirnova committed:
Merge branch '3.x' into PT-2340_Support_MySQL_8.4

chaleic commented:
@svetasmirnova , now that mysql 8.4 has been merged into the codebase, when might the next version of percona tools be built and the rpm files generated?   Thank you in advance

svetasmirnova commented:
@chaleic release is in progress, we actively working on it. Please follow Percona's website and blog for updates.
